pkg/users/account-create-dialog.js

```diff
@@ -0,0 +1,284 @@
+                        <div className="progress-bar" />
+                        <div className="progress-bar" />
+                    </div>
+                    <div>
+                        <span id="accounts-create-password-meter-message" className="help-block">{password_message}</span>
+                    </div>
+                </div>
+
+                <label className="control-label">{_("Access")}</label>
+                <label className="checkbox-inline">
+                    <input type="checkbox" id="accounts-create-locked"
+                     value={locked} onChange={event => change("locked", event.target.checked)} />
+                    <span>{_("Lock Account")}</span>
+                </label>
+            </form>
+        </div>);
+}
+
+function is_valid_char_username(c) {
+    return (c >= 'a' && c <= 'z') ||
+        (c >= 'A' && c <= 'Z') ||
+        (c >= '0' && c <= '9') ||
+        c == '.' || c == '_' || c == '-';
+}
+
+function validate_username(username, accounts) {
+    if (!username)
+        return _("No user name specified");
+
+    for (var i = 0; i < username.length; i++) {
+        if (!is_valid_char_username(username[i]))
+            return _("The user name can only consist of letters from a-z, digits, dots, dashes and underscores.");
+    }
+
+    for (var k = 0; k < accounts.length; k++) {
+        if (accounts[k].name == username)
+            return _("This user name already exists");
+    }
+
+    return null;
+}
+
+function suggest_username(realname) {
+    function remove_diacritics(str) {
+        var translate_table = {
+            a :  '[àáâãäå]',
+            ae:  'æ',
+            c :  '[čç]',
+            d :  'ď',
+            e :  '[èéêë]',
+            i :  '[íìïî]',
+            l :  '[ĺľ]',
+            n :  '[ňñ]',
+            o :  '[òóôõö]',
+            oe:  'œ',
+            r :  '[ŕř]',
+            s :  'š',
+            t :  'ť',
+            u :  '[ùúůûűü]',
+            y :  '[ýÿ]',
+            z :  'ž',
+        };
+        for (var i in translate_table)
+            str = str.replace(new RegExp(translate_table[i], 'g'), i);
+
+        for (var k = 0; k < str.length;) {
+            if (!is_valid_char_username(str[k]))
+                str = str.substr(0, k) + str.substr(k + 1);
+            else
+                k++;
+        }
+
+        return str;
+    }
+
+    var result = "";
+    var name = realname.split(' ');
+
+    if (name.length === 1)
+        result = name[0].toLowerCase();
+    else if (name.length > 1)
+        result = name[0][0].toLowerCase() + name[name.length - 1].toLowerCase();
+
+    return remove_diacritics(result);
+}
+
+export function account_create_dialog(accounts) {
+    let dlg = null;
+    const state = {
+        real_name: "",
+        user_name: "",
+        password: "",
+        password_confirm: "",
+        password_strength: "",
+        password_message: "",
+        locked: false
+    };
```

Comment:
I'd say it comes down to whether or not we use show_modal_dialog (or something similar) or 
`<Modal show={dialog_is_visible}>`.

I am in favor of show_modal_dialog (although I also want to replace it with something simpler and more general). I tried the other approach with the superuser buttons, and I didn't like it at all. Operating a modal dialog is more properly represented by calling a function than changing a prop on a component.

The `<Modal show={...}>` approach is very similar to how we handled dialogs with jQuery, and it was super messy to get the necessary parameters to a dialog. With React I guess we can just instantiate the same dialog class in all the places where it can be activated, hmm....

But, yes, it is worth trying harder with the dialogs.

Reply:
it's acually more like `state.add_dialog_visible && <Modal ...>` in the parent. A component shouldn't know by itself whether it's visible or not, that's the decision of the parent. But again, it's fine to not completely rewrite that jQuery-style "create a complex object" logic in one go :-)

Reply:
I don't see why you bring up jQuery, honestly.